akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala

```diff
@@ -136,52 +66,213 @@ abstract class PathDirectives extends MiscDirectives {
    * If the request path ends with a slash, redirect to the same uri without trailing slash in the path.
    *
    * '''Caveat''': [[pathSingleSlash]] directive will not match inside of this directive.
+   *
+   * @param redirectionType A status code from StatusCodes, which must be a redirection type.
    */
-  @varargs
-  def redirectToNoTrailingSlashIfPresent(redirectionStatusCode: StatusCode, innerRoute: Route, moreInnerRoutes: Route*): Route =
-    RedirectToNoTrailingSlashIfPresent(redirectionStatusCode)(innerRoute, moreInnerRoutes.toList)
-
-  private def RawPathPrefixForMatchers(matchers: immutable.Seq[PathMatcher[_]]): Directive =
-    Directives.custom(RouteStructure.RawPathPrefix(matchers))
+  def redirectToNoTrailingSlashIfPresent(redirectionType: StatusCode, inner: Supplier[Route]): Route = ScalaRoute {
+    redirectionType match {
+      case r: Redirection ⇒ D.redirectToNoTrailingSlashIfPresent(r) { inner.get().toScala }
+      case _              ⇒ throw new IllegalArgumentException("Not a valid redirection status code: " + redirectionType)
+    }
+  }
 
-  private def RawPathPrefixTestForMatchers(matchers: immutable.Seq[PathMatcher[_]]): Directive =
-    Directives.custom(RouteStructure.RawPathPrefixTest(matchers))
+  //------ extra java-specific methods
 
-  private def joinWithSlash(matchers: immutable.Seq[PathMatcher[_]]): immutable.Seq[PathMatcher[_]] = {
-    def join(result: immutable.Seq[PathMatcher[_]], next: PathMatcher[_]): immutable.Seq[PathMatcher[_]] =
-      result :+ PathMatchers.SLASH :+ next
+  // Java-specific since there's no Java API to create custom PathMatchers. And that's because there's no Path model in Java.
+  /**
+   * Consumes a leading slash and extracts the next path segment, unmarshalling it and passing the result to the inner function.
+   */
+  def pathPrefix[T](t: Unmarshaller[String, T], inner: java.util.function.Function[T, Route]): Route = ScalaRoute {
+    D.pathPrefix(PathMatchers.Segment)(unmarshal(t, inner))
+  }
 
-    matchers.foldLeft(immutable.Seq.empty[PathMatcher[_]])(join)
+  /**
+   * Consumes a leading slash and extracts the next path segment, unmarshalling it and passing the result to the inner function,
+   * expecting the full path to have been consumed then.
+   */
+  def path[T](t: Unmarshaller[String, T], inner: java.util.function.Function[T, Route]): Route = ScalaRoute {
+    D.path(PathMatchers.Segment)(unmarshal(t, inner))
   }
 
-  private def convertMatchers(matchers: Seq[AnyRef]): immutable.Seq[PathMatcher[_]] = {
-    def parse(matcher: AnyRef): PathMatcher[_] = matcher match {
-      case p: PathMatcher[_] ⇒ p
-      case name: String      ⇒ PathMatchers.segment(name)
-      case x                 ⇒ throw new IllegalArgumentException(s"Matcher of class ${x.getClass} is unsupported for PathDirectives")
-    }
+  private def unmarshal[T](t: Unmarshaller[String, T], inner: java.util.function.Function[T, Route]) = { element: String ⇒
+    D.extractRequestContext { ctx ⇒
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l263)